### PR TITLE
Add Operator Lab

### PR DESCRIPTION
**Pull request by dvob**

### CODE REVIEW

**Comment by madchr1st** [comment]
Looks  very good, thank you @dsbrng25b !

File: additional-labs/operators.md [hunk 1]
```diff
@@ -0,0 +1,168 @@
+# Operators
+
+Operators sind eine Art und Weise wie man Kubernetes-native Applikationen paketieren, deployen und verwalten kann. Kubernetes-native Applikationen sind Applikationen, die einerseits in Kubernetes/OpenShift deployed sind und andererseits auch über das Kubernetes/OpenShift-API (kubectl/oc) verwaltet werden. Seit OpenShift 4 verwendet auch OpenShift selber eine Reihe von Operators um den OpenShift-Cluster, also sich selber zu verwalten.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Operators sind eine Art und Weise wie man Kubernetes-native Applikationen paketieren, deployen und verwalten kann. Kubernetes-native Applikationen sind Applikationen, die einerseits in Kubernetes/OpenShift deployed sind und andererseits auch über das Kubernetes/OpenShift-API (kubectl/oc) verwaltet werden. Seit OpenShift 4 verwendet auch OpenShift selber eine Reihe von Operators um den OpenShift-Cluster, also sich selber, zu verwalten.
```

File: README.md [hunk 2]
```diff
@@ -33,6 +33,7 @@ __Ziele dieses Techlabs__:
 - [Cron Jobs in OpenShift](additional-labs/cronjobs_and_jobs.md)
 - [ConfigMaps](additional-labs/configmaps.md)
 - [Eigene OpenShift Entwicklungsumgebung](additional-labs/development_environment.md)
+- [Operators](additional-labs/operators.md)
```
Comment (bliemli) [inline README.md]:
```suggestion
- [Operators](additional-labs/operators.md)
- [Eigene OpenShift Entwicklungsumgebung](additional-labs/development_environment.md)
```

File: additional-labs/operators.md [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+## Einführung / Begriffe
+
+Um zu verstehen was ein Operator ist und wie er funktioniret, schauen wir zunächst das Konzept des Controllers an, da Operators auf dem Konzept der Controller basieren.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Um zu verstehen, was ein Operator ist und wie er funktioniert, schauen wir zunächst den sogenannten Controller an, da Operators auf dessen Konzept basieren.
```

File: additional-labs/operators.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+### Controller
+
+Ein Controller besteht aus einem Loop, in welchem immer wieder der gewünschte Zustand (_desired state_) und der aktuelle Zustand (_actual state/obseved state_) des Clusters gelesen werden. Wenn der aktuelle Zustand nicht dem gewünschten Zustand entspricht, versucht der Controller den gewünschten Zustand herzustellen. Der gewünschte Zustand wird mit Resourcen (Deployments, ReplicaSets, Pods, Services, etc.) beschrieben.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Ein Controller besteht aus einem Loop, in welchem immer wieder der gewünschte Zustand (_desired state_) und der aktuelle Zustand (_actual state/obseved state_) des Clusters gelesen werden. Wenn der aktuelle Zustand nicht dem gewünschten Zustand entspricht, versucht der Controller den gewünschten Zustand herzustellen. Der gewünschte Zustand wird mit Ressourcen (Deployments, ReplicaSets, Pods, Services, etc.) beschrieben.
```

File: additional-labs/operators.md [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+Die ganze Funktionsweise von OpenShift/Kubernetes basiert auf diesem Muster. Auf dem Master (controller-manager) laufen eine vielzahl von Controllern, welche auf Grund von Resourcen (ReplicaSets, Pods, Services, etc.) den gewünschten Zustand herstellen. Erstellt man z.B. ein ReplicaSet, sieht dies der ReplicaSet-Controller und erstellt als Folge die entsprechende Anzahl von Pods.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Die ganze Funktionsweise von OpenShift/Kubernetes basiert auf diesem Muster. Auf dem Master (controller-manager) laufen eine Vielzahl von Controllern, welche aufgrund von Ressourcen (ReplicaSets, Pods, Services, etc.) den gewünschten Zustand herstellen. Erstellt man z.B. ein ReplicaSet, sieht dies der ReplicaSet-Controller und erstellt als Folge die entsprechende Anzahl von Pods.
```

File: additional-labs/operators.md [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+__Optional__: Der Artikel [The Mechanics of Kubernetes](https://medium.com/@dominik.tornow/the-mechanics-of-kubernetes-ac8112eaa302) gibt einen tiefen Einblick in die Funktionsweise von Kubernetes. In der Grafik im Abschnitt _Cascading Commands_ wird schön aufgezeigt das vom Erstellen eines Deployments bis zum effektiven Starten der Pods vier verschiedene Controller involviert sind.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
__Optional__: Der Artikel [The Mechanics of Kubernetes](https://medium.com/@dominik.tornow/the-mechanics-of-kubernetes-ac8112eaa302) gibt einen tiefen Einblick in die Funktionsweise von Kubernetes. In der Grafik im Abschnitt _Cascading Commands_ wird schön aufgezeigt, dass vom Erstellen eines Deployments bis zum effektiven Starten der Pods vier verschiedene Controller involviert sind.
```

File: additional-labs/operators.md [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+### Operator
+
+Ein Operator ist ein Controller, welcher zuständig ist eine Applikation zu installieren und zu verwalten. Ein Operator hat also applikations-spezifisches Wissen. Dies ist insbesondere bei komplexeren Applikationen nützlich, welche aus verschiedenen Komponenten bestehen oder zusätzlichen Administrationsaufwand erfordern (z.B. neu gestartete Pods müssen zu einem Applikations-Cluster hinzugefügt werden, etc.).
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Ein Operator ist ein Controller, welcher dafür zuständig ist, eine Applikation zu installieren und zu verwalten. Ein Operator hat also applikations-spezifisches Wissen. Dies ist insbesondere bei komplexeren Applikationen nützlich, welche aus verschiedenen Komponenten bestehen oder zusätzlichen Administrationsaufwand erfordern (z.B. neu gestartete Pods müssen zu einem Applikations-Cluster hinzugefügt werden, etc.).
```

File: additional-labs/operators.md [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+Auch für den Operator muss der gewünschte Zustand durch eine Resource abgebildet werden. Dazu gibt es sogenannte Custom Resource Definitions (CRD). Mit CRDs kann man in Openshitf/Kubernetes beliebige neue Resourcen definieren. Der Operator schaut dann konstant (_watch_), ob Custom Resources verändert werden, für welche der Operator zuständig ist und führt entsprechend der Zielvorgabge in der Custom Resource Aktionen aus.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Auch für den Operator muss der gewünschte Zustand durch eine Ressource abgebildet werden. Dazu gibt es sogenannte Custom Resource Definitions (CRD). Mit CRDs kann man in OpenShitf/Kubernetes beliebige neue Ressourcen definieren. Der Operator schaut dann konstant (_watch_), ob Custom Resources verändert werden, für welche der Operator zuständig ist und führt entsprechend der Zielvorgabge in der Custom Resource Aktionen aus.
```

File: additional-labs/operators.md [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+Operators erleichtern es also komplexere Applikationen zu betreiben, da das Management vom Operator übernommen wird. Allfällige komplexe Konfigurationen werden durch Custom Resources des Operators abstrahiert und Betriebsaufgaben wie Backups oder rotieren von Zertifikaten etc. könne auch vom Operator ausgeführt werden.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Operators erleichtern es also komplexere Applikationen zu betreiben, da das Management vom Operator übernommen wird. Allfällige komplexe Konfigurationen werden durch Custom Resources abstrahiert und Betriebsaufgaben wie Backups oder das Rotieren von Zertifikaten etc. können auch vom Operator ausgeführt werden.
```

File: additional-labs/operators.md [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+
+## Installation eines Operators
+
+Ein Operator läuft wie eine normale Applikation als Pod im Cluster. Zur Installation eines Operators gehören in der Regel die folgenden Resourcen:
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Ein Operator läuft wie eine normale Applikation als Pod im Cluster. Zur Installation eines Operators gehören in der Regel die folgenden Ressourcen:
```

File: additional-labs/operators.md [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+* ***Custom Resource Definition***: Damit die neuen Custom Resources angelegt werden können, welche der Operator behandelt, müssen die entsprechenden CRDs installiert werden.
+* ***Service Account***: Ein Service Account mit welchem der Operator läuft.
+* ***Role und RoleBinding***: Mit einer Role definiert man alle Rechte, welche der Operator braucht. Dazu gehören mindestens Rechte auf die eigene Custom Resource. Mit einem RoleBinding wird die neue Role an dem Service Account des Operators zugewiesen.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
* ***Role und RoleBinding***: Mit einer Role definiert man alle Rechte, welche der Operator braucht. Dazu gehören mindestens Rechte auf die eigene Custom Resource. Mit einem RoleBinding wird die neue Role dem Service Account des Operators zugewiesen.
```

File: additional-labs/operators.md [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+* ***Deployment***: Ein Deployment um den eigentlichen Operator laufen zu lassen. Der Operator läuft meistens nur einmal (replicas auf Eins eingestellt), da sich sonst die verschiedenen Operator-Instanzen gegenseitig in die Quere kommen würden.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
* ***Deployment***: Ein Deployment um den eigentlichen Operator laufen zu lassen. Der Operator läuft meistens nur einmal (Replicas auf 1 eingestellt), da sich sonst die verschiedenen Operator-Instanzen gegenseitig in die Quere kommen würden.
```

File: additional-labs/operators.md [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+
+Zunächst legen wir ein neues Projekt an:
+
+```
+oc new-project my-etcd-cluster
+```
+
+Wir installieren den ETCD-Operator über die OpenShift WebConsole. Unter Operators -> OperatorHub finden wir den Katalog mit den verfügbaren Operators.
+
+* Nach `etcd` suchen, ETCD-Operator auswählen und Warnung, dass es sich um einen Community Operator handelt bestätigen.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
* Nach `etcd` suchen, ETCD-Operator auswählen und Warnung, dass es sich um einen Community Operator handelt, bestätigen.
```

File: additional-labs/operators.md [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+kind: EtcdCluster
+metadata:
+  name: example
+  namespace: my-etcd-cluster
+spec:
+  size: 3
+  version: 3.2.13
+```
+
+Nun könne wir beobachten, dass drei Pods für den ETCD-Cluster erstellt werden/wurden:
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Nun können wir beobachten, dass drei Pods für den ETCD-Cluster erstellt werden/wurden:
```

File: additional-labs/operators.md [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+etcdoperator.v0.9.4   etcd      0.9.4     etcdoperator.v0.9.2   Succeeded
+```
+
+Danach entfernen wir die ClusterServiceVersion:
+
+```
+oc delete csv etcdoperator.v0.9.4
+```
+
+Mit `oc get pod` können wir nun verifizieren das der Operator Pod entfernt wurde.
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
Mit `oc get pod` können wir nun verifizieren, dass der Operator Pod entfernt wurde.
```

File: additional-labs/operators.md [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+## Links
+
+* OpenShift Dokumentation über Operators: https://docs.openshift.com/container-platform/4.3/operators/olm-what-operators-are.html
+* Buch über Operators: https://www.redhat.com/cms/managed-files/cl-oreilly-kubernetes-operators-ebook-f21452-202001-en_2.pdf
```
Comment (bliemli) [inline additional-labs/operators.md]:
```suggestion
* [OpenShift Dokumentation zu Operators](https://docs.openshift.com/container-platform/4.3/operators/olm-what-operators-are.html)
* [Buch von O'Reilly über Operators](https://www.redhat.com/cms/managed-files/cl-oreilly-kubernetes-operators-ebook-f21452-202001-en_2.pdf)
```

**Review by bliemli — CHANGES_REQUESTED** [review]

**Review by bliemli — APPROVED** [review]

**Pull request merged**